lib/code_corps/github/api.ex

```diff
@@ -0,0 +1,34 @@
+defmodule CodeCorps.GitHub.API do
+  alias CodeCorps.{
+    GitHub,
+    GitHub.APIError,
+    GitHub.HTTPClientError
+  }
+
+  @spec request(GitHub.method, String.t, GitHub.headers, GitHub.body, list) :: GitHub.response
+  def request(method, url, headers, body, options) do
+    method
+    |> :hackney.request(url, headers, body, options)
+    |> marshall_response()
+  end
+
+  @typep http_success :: {:ok, integer, [{String.t, String.t}], String.t}
+  @typep http_failure :: {:error, term}
+
+  @spec marshall_response(http_success | http_failure) :: GitHub.response
+  defp marshall_response({:ok, status, _headers, body}) when status in 200..299 do
+    body |> Poison.decode
+  end
+  defp marshall_response({:ok, 404, _headers, body}) do
+    {:error, APIError.new({404, %{"message" => body}})}
+  end
+  defp marshall_response({:ok, status, _headers, body}) when status in 400..599 do
+    case body |> Poison.decode do
+      {:ok, json} -> {:error, APIError.new({status, json})}
+      {:error, error} -> {:error, error}
+    end
+  end
+  defp marshall_response({:error, reason}) do
+    {:error, HTTPClientError.new(reason: reason)}
+  end
+end
```

Comment:
This should be the only place where requests to github are actually made, and even then, we should keep it as flat as simple as possible. 

We need to write a test module for this file to, and it should be the only one to use bypass.